### PR TITLE
refactor(batch,agg): merge `Aggregator::update*` and refactor `SortAggExecutor`

### PR DESCRIPTION
**Pull request by richardchien**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

Before this PR,

- `HashAggExecutor` used `Aggregator::update_with_row` to feed the aggregator one row at a time and used `Aggregator::output` to generate the aggregation result
- `SortAggExecutor` used `Aggregator::update_and_output_with_sorted_groups` to feed the aggregator with multiple rows together with `EqGroups`, and used `Aggregator::output` to generate the aggregation result

Several problems existed:

- `Aggregator::update(input: &DataChunk)` was not used
- `Aggregator::update_and_output_with_sorted_groups` impls needed to know about `EqGroups`, and control the timing of output and reset of internal state, making the code unnecessarily complicated and inconsistent
- `Aggregator::output` did not clear the internal state of the aggregator, so it can't be called multiple times during feeding updates

Changes in this PR:

- Remove `Aggregator::update`
- Rename `Aggregator::update_with_row` to `Aggregator::update_single`
- Add `Aggregator::update_multi` to feed Aggregator a slice of input
- Remove `Aggregator::update_and_output_with_sorted_groups`, and move the control of output out of aggregators to `SortAggExecutor`
- Add `Aggregator::output_and_reset` to output the aggregated result and reset the internal state

After this PR, `Aggregator`s are intended to be used as follows:

```rust
agg.update_single(input, 0);
agg.update_multi(input, 1, input.len());
agg.output(builder); // output the current result, but don't clear internal state
agg.output(builder); // output the same result as the above
agg.output_and_reset(builder); // output the result and reset aggregator state
agg.update_multi(input2, 0, input2.len());
agg.update_multi(input3, 0, input3.len());
agg.output(builder); // output the agg result of input2 & input3
```

`Aggregator`s only need to take care about **how to aggregate values**, but never need to know the boundary of `EqGroups`.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

### CODE REVIEW

File: src/expr/src/vector_op/agg/approx_count_distinct.rs [hunk 1]
```diff
@@ -122,17 +121,25 @@ impl Aggregator for ApproxCountDistinct {
...
+        input: &DataChunk,
+        start_row_id: usize,
+        end_row_id: usize,
+    ) -> Result<()> {
         let array = input.column_at(self.input_col_idx).array_ref();
-        for datum_ref in array.iter() {
+        for datum_ref in array
+            .iter()
+            .skip(start_row_id)
+            .take(end_row_id - start_row_id)
```
Comment (skyzh) [inline src/expr/src/vector_op/agg/approx_count_distinct.rs:141]: .take will stop when we reach the end of the array? Should assert start_row_id and end_row_id in range.

**Review by skyzh — APPROVED** [review]
Rest LGTM

File: src/expr/src/vector_op/agg/approx_count_distinct.rs (hunk 1 above)
Comment (richardchien) [reply src/expr/src/vector_op/agg/approx_count_distinct.rs:141]: I think the cases where `start_row_id` and `end_row_id` are mis-specified to be out of input data chunk is rare. More possible is to call `update_multi` with overlapped ranges, which can't be detected in `update_multi` impls. Adding asserts to `start/end_row_id` here doesn't help a lot and makes code ugly.🤔

Comment (skyzh) [reply src/expr/src/vector_op/agg/approx_count_distinct.rs:141]: Yes, I'd agree on that. I'm just thinking we could add asserts to make things easier if there's really a bug in the future. Maybe `debug_assert!` which will only take effect in debug mode. Anyway, it looks good to me overall if we already ensure everything is in range before calling this function.

**Pull request merged**